tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -0,0 +1,1414 @@
+   delete[] recordFooter;
+
+   RNTupleDescriptorBuilder descBuilder;
+   descBuilder.SetFromHeader(header);
+   descBuilder.AddClustersFromFooter(footer);
+   delete[] header;
+   delete[] footer;
+   return descBuilder.MoveDescriptor();
+}
+
+
+ROOT::Experimental::RNTupleDescriptor ROOT::Experimental::Detail::RPageSourceFile::DoAttach()
+{
+   char ident[4];
+   Read(ident, 4, 0);
+   if (std::string(ident, 4) == "root")
+      return AttachTFile();
+
+   return AttachRaw();
+}
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::PopulatePageFromCluster(
+   ColumnHandle_t columnHandle, const RClusterDescriptor &clusterDescriptor, ClusterSize_t::ValueType clusterIndex)
+{
+   auto columnId = columnHandle.fId;
+   auto clusterId = clusterDescriptor.GetId();
+   const auto &pageRange = clusterDescriptor.GetPageRange(columnId);
+
+   // TODO(jblomer): binary search
+   RClusterDescriptor::RPageRange::RPageInfo pageInfo;
+   decltype(clusterIndex) firstInPage = 0;
+   for (const auto &pi : pageRange.fPageInfos) {
+      if (firstInPage + pi.fNElements > clusterIndex) {
+         pageInfo = pi;
+         break;
+      }
+      firstInPage += pi.fNElements;
+   }
+   R__ASSERT(firstInPage <= clusterIndex);
+   R__ASSERT((firstInPage + pageInfo.fNElements) > clusterIndex);
+
+   auto element = columnHandle.fColumn->GetElement();
+   auto elementSize = element->GetSize();
+
+   auto pageSize = pageInfo.fLocator.fBytesOnStorage;
+   void *pageBuffer = malloc(std::max(pageSize, static_cast<std::uint32_t>(elementSize * pageInfo.fNElements)));
+   R__ASSERT(pageBuffer);
+   Read(pageBuffer, pageSize, pageInfo.fLocator.fPosition);
+
+   auto bytesOnStorage = (element->GetBitsOnStorage() * pageInfo.fNElements + 7) / 8;
+   if (pageSize != bytesOnStorage) {
+      fDecompressor(pageBuffer, pageSize, bytesOnStorage);
+      pageSize = bytesOnStorage;
+   }
+
+   if (!element->IsMappable()) {
+      pageSize = elementSize * pageInfo.fNElements;
+      auto unpackedBuffer = reinterpret_cast<unsigned char *>(malloc(pageSize));
+      R__ASSERT(unpackedBuffer != nullptr);
+      element->Unpack(unpackedBuffer, pageBuffer, pageInfo.fNElements);
+      free(pageBuffer);
+      pageBuffer = unpackedBuffer;
+   }
+
+   auto indexOffset = clusterDescriptor.GetColumnRange(columnId).fFirstElementIndex;
+   auto newPage = fPageAllocator->NewPage(columnId, pageBuffer, elementSize, pageInfo.fNElements);
+   newPage.SetWindow(indexOffset + firstInPage, RPage::RClusterInfo(clusterId, indexOffset));
+   fPagePool->RegisterPage(newPage,
+      RPageDeleter([](const RPage &page, void */*userData*/)
+      {
+         RPageAllocatorFile::DeletePage(page);
+      }, nullptr));
+   return newPage;
+}
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageSourceFile::PopulatePage(
+   ColumnHandle_t columnHandle, NTupleSize_t globalIndex)
+{
+   auto columnId = columnHandle.fId;
```

Comment:
Here and in the surrounding functions, some variables are ```const auto``` and some are just ```auto```.   It might help readability to mark all the variable that are not intended to be modified (i.e. used as accelerator/short-names) to all be marked const (it appears that const is used currently only for non-numerical read-only variable).